File: experiments/examples/service/routers/images.py
# Copyright 2024 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from fastapi import APIRouter, HTTPException
from models.common_model import ImageRequest, TextValue, ProductAttributes

from google.cloud.ml.applied.images import image_to_text

router = APIRouter()


@router.post("/api/v1/genai/images/attributes")
def generate_attributes(image_request: ImageRequest) -> ProductAttributes:
    """
    Extracts attributes detected from product's image.
    """
    try:
        response = image_to_text.image_to_attributes(image_request)
    except Exception as e:
        print(f"ERROR: Image Attribute Generation Error: -> {e}")
        raise HTTPException(status_code=500, detail=str(e))
    else:
        return response


@router.post("/api/v1/genai/images/descriptions")
def generate_description(image_request: ImageRequest) -> TextValue:
    """
    Generates a product description from an image of product.
    """
    try:
        response = image_to_text.image_to_product_description(image_request.image)
    except Exception as e:
        print(f"ERROR: Image Description Generation Error: -> {e}")
        raise HTTPException(status_code=500, detail=str(e))
    else:
        return response

File: demos/digital-commerce/apps/api/src/routes/images.ts
// Copyright 2024 Google, LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import {Request, Response, Router} from 'express';
import {api} from 'model';
import sessionManager from '../state';
import {utils} from 'model';

import { generateFailedDependencyResponse, extractTextCandidates} from '../utils';
import { GenerativeContentBlob, InlineDataPart } from '@google/generative-ai';

const router = Router();

/**
 * Handle image to Category[]
 */
router.post('/', (req: Request, resp: Response) => {
    const imagePrompt = req.body as api.ImagePromptRequest;
    const sessionID = imagePrompt.sessionID;
    const generativeSession = sessionManager.getSession(sessionID);
    if (generativeSession) {
        const model = generativeSession.model;
        model
            .generateContent({
                contents: [
                    {
                        role: 'user',
                        parts: [
                            {text: imagePrompt.prompt},
                            ...imagePrompt.value.map(
                                (image) =>
                                    ({
                                        inlineData: {
                                            data: utils.stripBase64Prefix(image.base64),
                                            mimeType: image.type,
                                        } as GenerativeContentBlob,
                                    } as InlineDataPart)
                            ),
                        ],
                    },
                ],
            })
            .then((result) => {
                // console.log(`SUCCESS: ${extractTextCandidates(result)}`)
                resp.status(200).send(extractTextCandidates(result));
            });
    } else {
        console.error(`ERROR ...${JSON.stringify(resp)}`)
        generateFailedDependencyResponse(resp);
    }
});

export default router;
